compiler/semantic/expr.go

```diff
@@ -374,14 +398,32 @@ func semAssignment(scope *Scope, a ast.Assignment) (dag.Assignment, error) {
 		lhs = &dag.This{"This", []string{name}}
 	} else if agg, ok := a.RHS.(*ast.Agg); ok {
 		lhs = &dag.This{"This", []string{agg.Name}}
-	} else if isThis(a.RHS) {
-		return dag.Assignment{}, errors.New(`cannot assign to "this"`)
+	} else if v, ok := rhs.(*dag.Var); ok {
+		lhs = &dag.This{"This", []string{v.Name}}
 	} else {
-		lhs, err = semField(scope, a.RHS, true)
+		lhs, err = semExpr(scope, a.RHS)
 		if err != nil {
 			return dag.Assignment{}, errors.New("assignment name could not be inferred from rhs expression")
 		}
 	}
+	if summarize {
+		// Summarize always outputs its results as new records of "this"
+		// this so if we have an "as" that overrides "this", we just
+		// convert it back to a local this.
+		if dot, ok := lhs.(*dag.Dot); ok {
+			if v, ok := dot.LHS.(*dag.Var); ok && v.Name == "this" {
+				lhs = &dag.This{Kind: "This", Path: []string{dot.RHS}}
+			}
+		}
+	}
+	// Make sure we have a valid l-value for lhs.
```

Comment:
Nit: Existing comments say "lval" instead of "l-value".
```suggestion
	// Make sure we have a valid lval for lhs.
```